exporter/otelarrowexporter/metadata.go

```diff
@@ -0,0 +1,192 @@
+	lock sync.Mutex
+	size int
+}
+
+var _ exp = (*metadataExporter)(nil)
+
+func newMetadataExporter(cfg component.Config, set exporter.Settings, streamClientFactory streamClientFactory) (exp, error) {
+	oCfg := cfg.(*Config)
+	// use lower-case, to be consistent with http/2 headers.
+	mks := make([]string, len(oCfg.MetadataKeys))
+	for i, k := range oCfg.MetadataKeys {
+		mks[i] = strings.ToLower(k)
+	}
+	sort.Strings(mks)
+	if len(mks) == 0 {
+		return newExporter(cfg, set, streamClientFactory)
+	}
+	return &metadataExporter{
+		config:       oCfg,
+		settings:     set,
+		scf:          streamClientFactory,
+		metadataKeys: mks,
+	}, nil
+}
+
+func (e *metadataExporter) getSettings() exporter.Settings {
+	return e.settings
+}
+
+func (e *metadataExporter) getConfig() component.Config {
+	return e.config
+}
+
+func (e *metadataExporter) helperOptions() []exporterhelper.Option {
+	return []exporterhelper.Option{
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		exporterhelper.WithTimeout(e.config.TimeoutSettings),
+		exporterhelper.WithRetry(e.config.RetryConfig),
+		exporterhelper.WithQueue(e.config.QueueSettings),
+		exporterhelper.WithStart(e.start),
+		exporterhelper.WithShutdown(e.shutdown),
+	}
+}
+
+func (e *metadataExporter) start(_ context.Context, host component.Host) (err error) {
+	e.host = host
+	return nil
+}
+
+func (e *metadataExporter) shutdown(ctx context.Context) error {
+	var err error
+	e.exporters.Range(func(_ any, value any) bool {
+		be := value.(exp)
+		err = multierr.Append(err, be.shutdown(ctx))
+		return true
+	})
+	return err
+}
+
+func (e *metadataExporter) pushTraces(ctx context.Context, td ptrace.Traces) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+	return be.pushTraces(ctx, td)
+}
+
+func (e *metadataExporter) pushMetrics(ctx context.Context, md pmetric.Metrics) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+
+	return be.pushMetrics(ctx, md)
+}
+
+func (e *metadataExporter) pushLogs(ctx context.Context, ld plog.Logs) error {
+	s, mdata := e.getAttrSet(ctx, e.metadataKeys)
+
+	be, err := e.getOrCreateExporter(ctx, s, mdata)
+	if err != nil {
+		return err
+	}
+
+	return be.pushLogs(ctx, ld)
+}
+
+func (e *metadataExporter) getOrCreateExporter(ctx context.Context, s attribute.Set, md metadata.MD) (exp, error) {
+	v, ok := e.exporters.Load(s)
+	if ok {
+		return v.(exp), nil
+	}
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if e.config.MetadataCardinalityLimit != 0 && e.size >= int(e.config.MetadataCardinalityLimit) {
+		return nil, errTooManyExporters
+	}
+
+	newExp, err := newExporter(e.config, e.settings, e.scf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create exporter: %w", err)
+	}
+
+	var loaded bool
+	v, loaded = e.exporters.LoadOrStore(s, newExp)
```

Comment:
I'd like to make the call to `newExporter()` cost less, to keep this code-path efficient. The stuff in otelarrow.go lines 85-101 looks like it could move into the metadata exporter, so that at this point in the code, only a single object is allocated. I believe the netReporter and userAgent values can be calculated once, not once per exporter.

Reply:
Sounds good, fixed

Reply:
I think moving the netreporter uncovered a race where e.TraceExporter seems to have a non deterministic value, sometimes it's nil and sometimes `ptraceotlp.NewGRPCClient(e.clientConn)` is valid - can't figure out if its getting overwritten or if it takes a while for e.start to set the value. Will need to investigate more on monday

Reply:
This took forever to figure out because the test errors were so misleading - the main race that I fixed was a lack of synchronization between `e.exporters.Load(s)` and checking `e.size` in metadata.go.  They both need to be guarded by the mutex otherwise e.size will be wrong if a thread Loads an empty value while another thread is creating an exporter for the same attribute key. Then both threads will increment e.size despite only one exporter being stored for the key.

Reply:
Makes sense. I re-reviewed the logic around .size, thanks for the improvements.